### PR TITLE
Use vscode watches for tsserver

### PR DESCRIPTION
**Pull request by sheetalkamat**
Works with https://github.com/microsoft/TypeScript/pull/54662

### CODE REVIEW

**Comment by bpasero** [comment]
🆒 , is this the change that would enable file watching in all of TS extension with the VS Code file watcher or is there more needed?

File: extensions/typescript-language-features/src/typescriptServiceClient.ts [hunk 1]
```diff
@@ -973,6 +975,45 @@ export default class TypeScriptServiceClient extends Disposable implements IType
 			case EventName.projectLoadingFinish:
 				this.loadingIndicator.finishedLoadingProject((event as Proto.ProjectLoadingFinishEvent).body.projectName);
 				break;
+
+			case EventName.createDirectoryWatcher:
+				this.createFileSystemWatcher(event.body.id, new vscode.RelativePattern(vscode.Uri.file(event.body.path), event.body.recursive ? '**' : '*'));
```
Comment (bpasero) [inline extensions/typescript-language-features/src/typescriptServiceClient.ts:980]: Note that this should work fine for non-recursive watchers (which are always preferred) but has some limitations for recursive watchers if the path is inside the workspace. Please review: https://code.visualstudio.com/api/references/vscode-api#workspace.createFileSystemWatcher

**Comment by sheetalkamat** [comment]
> is this the change that would enable file watching in all of TS extension with the VS Code file watcher or is there more needed?

Yes https://github.com/microsoft/TypeScript/pull/54662 should redirect all the watch calls by tsserver into vscode if invoked with `--canUseWatchEvents`

File: extensions/typescript-language-features/src/typescriptServiceClient.ts (hunk 1 above)
Comment (sheetalkamat) [reply extensions/typescript-language-features/src/typescriptServiceClient.ts:980]: I think I would need some help in fixing this part then. I think we would want to watch the folder if its workspace folder or outside of workspace otherwise watch workspace and only redirect to tsserver when events it is interested in are generated (inside the sub path area)
Does this need to handle more workspace folders added after watching has started etc. I am not sure how it works with all that but if someone more expert can take over that would be great.

Comment (bpasero) [reply extensions/typescript-language-features/src/typescriptServiceClient.ts:980]: @sheetalkamat 

> I think I would need some help in fixing this part then.

Just to clarify: you are free to `createFileSystemWatcher` on any path from your extension, both recursive and non-recursive. However, one limitation today is that recursive watching inside the workspace is superseded by the watcher that VS Code itself sets up when you open the workspace. That means, `files.watcherExclude` setting is applied and by default `**/node_modules/*/**` will be ignored. So if you need to watch recursively inside `node_modules` folders from your extension, we need to think how to support that, either by VS Code making this possible or you changing to non-recursive watching if possible. So please let me know if that is the case.

> Does this need to handle more workspace folders added after watching has started etc

Both recursive and non-recursive file watching can handle the case of new folders being added. But if your extension needs to create new file system watchers for new workspaces added, then you need to handle that case from your extension by listening to respective workspace folder change events. Maybe I misunderstood the question, can you rephrase?

Comment (sheetalkamat) [reply extensions/typescript-language-features/src/typescriptServiceClient.ts:980]: We do need to watch node_modules and we watch them recursively either to watch when program references files from it (we prefer that over watching individual files) and also to ensure that resolutions can be updated eg when npm install happens.

Comment (bpasero) [reply extensions/typescript-language-features/src/typescriptServiceClient.ts:980]: I am not sure I fully understand. Can you give examples of file changes in `node_modules` folder that you are interested in specifically?

**Comment by bpasero** [comment]
Fyi https://github.com/microsoft/vscode/pull/194776 landed which enables extensions to do full recursive file watching anywhere even within the workspace with custom exclude rules. It is behind proposed API (`createFileSystemWatcher`).

File: extensions/typescript-language-features/src/typescriptServiceClient.ts [hunk 2]
```diff
@@ -973,6 +975,46 @@ export default class TypeScriptServiceClient extends Disposable implements IType
 			case EventName.projectLoadingFinish:
 				this.loadingIndicator.finishedLoadingProject((event as Proto.ProjectLoadingFinishEvent).body.projectName);
 				break;
+
+			case EventName.createDirectoryWatcher:
+				this.createFileSystemWatcher(event.body.id, new vscode.RelativePattern(vscode.Uri.file(event.body.path), event.body.recursive ? '**' : '*'), [ /* TODO need to fill in excludes list */]);
```
Comment (bpasero) [inline extensions/typescript-language-features/src/typescriptServiceClient.ts:980]: @sheetalkamat fyi this is still required to make it work: we need to know what things to exclude, otherwise VS Code will refuse to accept this request for watching.

Comment (sheetalkamat) [reply extensions/typescript-language-features/src/typescriptServiceClient.ts:980]: I am not sure how to write exclude pattern here. We know things that should definitely be included (js, mjs. cjs, ts, mts, cts json and any dts files which is .d.someExtension) There is also capability to support extra extensions and that list is given by `protocol.ConfigureRequestArguments` so those are also something we care.

Comment (sheetalkamat) [reply extensions/typescript-language-features/src/typescriptServiceClient.ts:980]: But here is the list of things we definitely ignore..  If the file path has `/node_modules/.`,  `/.git` or `/.#` we definitely ignore. 
Will adding that pattern be enough instead

Comment (sheetalkamat) [reply extensions/typescript-language-features/src/typescriptServiceClient.ts:980]: Also if the directory watcher is not recursive, do we need exclude pattern and new API usage ?

Comment (bpasero) [reply extensions/typescript-language-features/src/typescriptServiceClient.ts:980]: @sheetalkamat yes, you are right, only **recursive** watchers require `exclude: string[]` to leverage the new capability, you do not have to define excludes for non-recursive watchers.

For the recursive watcher we actually get some nice extended glob capabilities because the glob pattern is directly interpreted in the native watcher and that one uses `micromatch`. With `micromatch` you can prefix patterns with `!` to negate them, so you can essentially do `excludes: ["!**/*.ts"]` to only include `*.ts` files. I am quite sure you can build a nice pattern for the files you are interested in with that approach.

However, another approach to above would be to not use `*` or `**` as pattern for the `createFileSystemWatcher(pattern)`. You can specify a glob pattern there as well to directly indicate the files you want to be notified about. Think about it like the counterpart of `exclude: string[]`, but for includes.

Note that glob patterns used in `pattern` of `createFileSystemWatcher(pattern)` and as `exclude: string[]` for non-recursive watchers are slightly less powerful. Specifically negates are not supported. See https://code.visualstudio.com/docs/editor/glob-patterns for reference.

Comment (bpasero) [reply extensions/typescript-language-features/src/typescriptServiceClient.ts:980]: Here is my suggestion:
* recursive watching
  * define `excludes` leveraging the [extended capabilities](https://github.com/micromatch/micromatch#matching-features) from `micromatch` with leading `!`
  * instead of `*`, pass in a pattern for the files you care about following our [glob reference](https://code.visualstudio.com/docs/editor/glob-patterns)
* non-recursive watching
  * leave `excludes` as `undefined`
  * instead of `*`, pass in a pattern for the files you care about following our [glob reference](https://code.visualstudio.com/docs/editor/glob-patterns)

**Comment by mjbvz** [comment]
@sheetalkamat I pushed a change that gates this feature behind a setting and also tries to clean up file watchers when the server restarts

In my testing, I see errors such as this in the developer tools:

```
 MainThreadFileSystemEventService#$watch(): failed to stat a resource for file watching (extension: vscode.typescript-language-features, path: file:///Users/matb/projects/node_modules, recursive: true, session: 0.16141590890668556): EntryNotFound (FileSystemError): Error: ENOENT: no such file or directory, stat '/Users/matb/projects/node_modules'
 ```
 
 Will this cause any issues for TypeScript? It seems like we may be trying to watch the contents of a directory that does not exist. Will we still pick up if the directory is created later?

**Comment by sheetalkamat** [comment]
> Will this cause any issues for TypeScript? It seems like we may be trying to watch the contents of a directory that does not exist. Will we still pick up if the directory is created later?

We do need to watch even if directory does not exist and get events when it gets created. (Without this we wouldnt be picking up changes to project resolutions because of `npm install`) 
@bpasero is this supported. Are the errors because i am yet to add exclude pattern.

File: extensions/typescript-language-features/src/typescriptServiceClient.ts (hunk 2 above)
Comment (sheetalkamat) [reply extensions/typescript-language-features/src/typescriptServiceClient.ts:990]: I thought about this. but there is no safe way to exclude any extension especially because we have language plugins that can handle the additional file extensions without informing typescript about it. So the exclude pattern would have to be paths of certain ignored patterns as mentioned in https://github.com/microsoft/vscode/pull/193848#discussion_r1357164773

Comment (bpasero) [reply extensions/typescript-language-features/src/typescriptServiceClient.ts:990]: Yeah I think you have to come up with the list that works best for you, but the list cannot be empty currently so that you opt-in to the new behaviour.

**Comment by bpasero** [comment]
@sheetalkamat no, that is not supported. you have to watch a parent folder that exists. Maybe then you will have to watch the workspace folder(s) and do a negated exclude pattern such as `!**/node_modules/**` to recursively watch any node module folder that is either there already or gets created.

**Comment by bpasero** [comment]
https://github.com/microsoft/vscode/pull/207145 implements support in our file watcher to watch paths that do not exist. This supports paths that do not exist in the beginning or are deleted and recreated later. I am actually using `fs.watchFile` for this purpose because I found this to be the only watcher that works with non-existing paths, since it is using polling.

File: extensions/typescript-language-features/src/typescriptServiceClient.ts [hunk 3]
```diff
@@ -973,6 +985,64 @@ export default class TypeScriptServiceClient extends Disposable implements IType
...
+
+		disposable.add(watcher.onDidChange(changeFile =>
+			this.executeWithoutWaitingForResponse('watchChange', { id, path: changeFile.fsPath, eventType: 'update' } satisfies Proto.WatchChangeRequestArgs)
+		));
+		disposable.add(watcher.onDidCreate(createFile =>
+			this.executeWithoutWaitingForResponse('watchChange', { id, path: createFile.fsPath, eventType: 'create' } satisfies Proto.WatchChangeRequestArgs)
+		));
+		disposable.add(watcher.onDidDelete(deletedFile =>
+			this.executeWithoutWaitingForResponse('watchChange', { id, path: deletedFile.fsPath, eventType: 'delete' } satisfies Proto.WatchChangeRequestArgs)
+		));
```
Comment (bpasero) [inline extensions/typescript-language-features/src/typescriptServiceClient.ts:1031]: @sheetalkamat I wonder if the communication between client and server should work in a way that we can collect multiple file events and send them over in batches. Here we seem to be doing 1 IPC roundtrip per single file event, but in reality there will be hundreds of events. 

Maybe that could go together with a utility to aggregate over multiple events for when we deal with changes to `node_modules` where you are only interested in the top level module event.

File: extensions/typescript-language-features/src/typescriptServiceClient.ts [hunk 4]
```diff
@@ -973,6 +993,120 @@ export default class TypeScriptServiceClient extends Disposable implements IType
...
+			if (event?.[typeOfEventToRemove]?.delete(path) && event[typeOfEventToRemove].size === 0) {
+				event[typeOfEventToRemove] = undefined;
+			}
+		};
+		const aggregateEvent = () => {
+			if (!event) {
+				this.watchEvents.set(id, event = {});
+			}
+			(event[eventType] ??= new Set()).add(path);
+		};
```
Comment (bpasero) [inline extensions/typescript-language-features/src/typescriptServiceClient.ts:1052]: Would it be more efficient to have these methods on the class instead of defining them over and over again for each event?

Comment (bpasero) [reply extensions/typescript-language-features/src/typescriptServiceClient.ts:1052]: Fyi we do something very similar before emitting events to the outside in core:

https://github.com/microsoft/vscode/blob/f0ce480524258473372e0a2e9a07af6f64526010/src/vs/platform/files/common/watcher.ts#L338

But we do this only per watcher instance and not across all instances.

Comment (sheetalkamat) [reply extensions/typescript-language-features/src/typescriptServiceClient.ts:1052]: In our case we have found that classes are slower (esp if properties are on prototype instead of instance) and we prefer this style of code. But if you prefer classes, feel free to change.

**Review by bpasero — DISMISSED** [review]

**Review by bpasero — APPROVED** [review]

**Review by Tyriar — APPROVED** [review]
